Added src to announcements

anirudh723 opened the pull request:
Added an image_src column to the database, and a corresponding imageSrc field to the Announcements DTO, and the create + get announcement requests. Note: the imageSrc is _optional_ for the create announcement request, since the user does not need to specify an image for an announcement. 

## Why

This will now enable users to upload/view announcements with images.

What benefit does this bring to the end user? Or, what benefit does this bring to developers 
working in the codebase?

Users will now get a better idea of an announcement with an image describing it.

## This PR

Added a sql migration file to add an image_src (string) column to our announcements table. Then, I added a string src field to the announcements DTO, matching the new table. I added an imageSrc field to the `PostAnnouncementRequest` object, which can be null. 

## Screenshots

Creating an announcement _with_ an image:
![image](https://user-images.githubusercontent.com/46979800/108939033-139c3480-761f-11eb-8118-bd96d8e1cfa3.png)

Creating an announcement _without_ an image:
![image](https://user-images.githubusercontent.com/46979800/108939072-2151ba00-761f-11eb-8404-c8184d59ca86.png)

Getting announcement (now you can see the newly added imageSrc field):
![image](https://user-images.githubusercontent.com/46979800/108939116-33cbf380-761f-11eb-8e7e-9709fe84c1df.png)

## Verification Steps

What steps did you take to verify your changes work? These should be clear enough for someone to be 
able to clone the branch and follow the steps themselves. 

I added announcements to the table with and without an imageSrc value. I added JUnit tests to create announcements (event specific and non event-specific) with images. I also added JUnit tests to get announcements that had images and also that didn't. Furthermore, I made test requests using Postman to verify the correct functionality.

liammoyn reviewed (commented):
We should call this column something more intuitive like 'image_src'.

Also, how does this ticket connect to the work @rymaju was talking about with actually uploading the image?

jackblanc commented on persist/src/main/resources/db/migration/V12__UpdateAnnouncements.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE announcements ADD COLUMN src VARCHAR DEFAULT null;
```
This should be formatted on multiple lines

jackblanc reviewed (commented):
This looks good, but where is the work for actually uploading images on the create announcement request going to happen? cc @rymaju

jackblanc commented on persist/src/main/resources/db/migration/V12__UpdateAnnouncements.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE announcements
+ADD COLUMN image_src VARCHAR DEFAULT null;
```
Don't you have to specify length for VARCHAR?

anirudh723 replied:
Good question. You don't _have_ to. In fact, some of our migration files don't. There's basically no difference in operation  performance between text, varchar(n), and varchar. It can be good to have a limit if we want to use it as a data constraint, but I assumed that we were being flexible with our data. 

If we want to add a limit, I can go ahead and do that. However, as you may know, url lengths can vary a lot, so we probably want to put a high cap (in which I don't think we should even do it), or will later have to add another migration file to update it if we come across an issue of it being too low.

jackblanc commented on persist/src/main/resources/db/migration/V12__UpdateAnnouncements.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE announcements
```
I'm going to merge another migration first so you have to update this to V13

anirudh723 commented:
@jackblanc I added the work for creating announcements with images (an image being optional of course), and added some screenshots in this PR showing that + getting announcements. Could you take a look at it when you have the chance? Thanks in advance!

rymaju commented on api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java:
```diff
@@ -3,27 +3,42 @@
+   * @param imageSrc image src url of the post request object to be created
+   */
+  public PostAnnouncementRequest(String title, String description, String imageSrc) {
     this.title = title;
```
If title and description are required should we check for non-null? Not sure if other requests do that too.

anirudh723 replied:
Hmm good question. I'm going to test that, because I believe we get back a malformed json response. I think since our db expects a non-null title and description, it might be catching that when trying to store a record.

rymaju commented on api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java:
```diff
@@ -44,6 +59,15 @@ public String getDescription() {
+   * @return an optional image src of the announcement in this post request object.
+   */
+  public Optional<String> getImageSrc() {
+    return imageSrc != null ? Optional.of(imageSrc) : Optional.empty();
```
Awesome I love optionals ❤️

anirudh723 replied:
Ooh good point. I will use the `isEmpty(String s)` method that checks for null and empty string 👍

rymaju commented on persist/src/main/resources/db/migration/V13__UpdateAnnouncements.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE announcements
+ADD COLUMN image_src VARCHAR DEFAULT null;
```
Should we put VARCHAR(<amount>)? like VARCHAR(255) or something? Not super big on SQL, any thoughts @liammoyn @jackblanc ?

anirudh723 replied:
Jack had a similar [question](https://github.com/Code-4-Community/lucys-love-bus-backend/pull/108#discussion_r581569071), and I gave my two cents on it

rymaju commented on service/src/main/java/com/codeforcommunity/processor/AnnouncementsProcessorImpl.java:
```diff
@@ -155,6 +156,9 @@ private AnnouncementsRecord announcementRequestToRecord(PostAnnouncementRequest
     AnnouncementsRecord newRecord = db.newRecord(ANNOUNCEMENTS);
```
Ideally we would like to take a BLOB as input (base64 string, I can work with you on this) and then push it up when we get the request. This will be as simple and copy pasting some code I've already written for the signup profile pic 😄

anirudh723 replied:
Yup

rymaju reviewed (commented):
I wanna wait on actual image uploading before we move forward with this PR. Overall it looks great though, lets find a time to pair in the next week. We can talk about it during the meeting!

anirudh723 commented on service/src/main/java/com/codeforcommunity/processor/AnnouncementsProcessorImpl.java:
```diff
@@ -163,7 +166,11 @@ private AnnouncementsRecord announcementRequestToRecord(PostAnnouncementRequest
     newRecord.setDescription(request.getDescription());
     if (request.getImageSrc().isPresent()) {
-      newRecord.setImageSrc(request.getImageSrc().get());
+      String filename = "annoucement-" + UUID.randomUUID();
```
nit: "annoucement" => "announcement"

anirudh723 replied:
But I will make this fix. Thanks for adding this snippet!

jackblanc commented on service/src/test/java/com/codeforcommunity/processor/AnnouncementsProcessorImplTest.java:
```diff
@@ -161,6 +228,53 @@ public void testPostAnnouncements2() {
 
+  // test posting an announcement with an image, without admin privileges
+  @Test
+  public void testPostAnnouncementsWithImageSrc1() {
```
I think it makes more sense to give descriptive names to the tests here

jackblanc reviewed (commented):
Looks great! Left a few comments.

anirudh723 commented on persist/src/main/resources/db/migration/V15__UpdateAnnouncements.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE announcements
+ADD COLUMN image_src VARCHAR DEFAULT null;
```
Don't have to, but I will do it

anirudh723 commented on service/src/main/java/com/codeforcommunity/processor/AnnouncementsProcessorImpl.java:
```diff
@@ -119,7 +123,13 @@ public PostAnnouncementResponse postEventSpecificAnnouncement(
         });
 
-    return announcementPojoToResponse(newAnnouncementsRecord.into(Announcements.class));
+    // the timestamp wasn't showing correctly, so just
+    // get the announcement directly from the database
```
Disregard this. I put it back to what it originally was

anirudh723 commented on service/src/test/java/com/codeforcommunity/processor/AnnouncementsProcessorImplTest.java:
```diff
@@ -118,6 +119,72 @@ public void testGetAnnouncements3() {
+    announcement1.setTitle("the first announcement title");
+    announcement1.setCreated(new Timestamp(START_TIMESTAMP_TEST));
+    announcement1.setDescription("the first announcement description");
+    announcement1.setImageSrc("https://facts.net/wp-content/uploads/2020/07/monarch-butterfly-facts.jpg");
```
Yeah I'm still working on these tests. Will push the fixed tests in my next-ish commit

jackblanc commented on service/src/test/java/com/codeforcommunity/processor/AnnouncementsProcessorImplTest.java:
```diff
@@ -121,6 +150,76 @@ public void testGetAnnouncements3() {
+    assertEquals(res.getAnnouncements().get(0).getId(), 0);
+    assertEquals(
+        res.getAnnouncements().get(0).getImageSrc(),
+        "https://facts.net/wp-content/uploads/2020/07/monarch-butterfly-facts.jpg");
```
Shouldn't this contain `announcements-uuid`?

jackblanc replied:
ok got it, thanks

jackblanc reviewed (commented):
I think this looks good, are announcements resolved? How are we testing the image?
